### PR TITLE
Add mg::gl::Texture abstraction for platform buffers

### PR DESCRIPTION
**Pull request by RAOF**
Unlike the existing platforms, the `eglstream-kms` platform can *only* provide us with a texture, not a thing that can be uploaded to a texture. Furthermore, it requires a different fragment shader (as, for that matter, do non-RGB(A) buffers in the `mesa-kms` platform).

This makes the existing `mir::renderer::gl::TextureSource` interface inappropriate for `eglstream-kms`.

Fortunately, it's easy enough for *all* the platforms to provide a GL texture rather than a thing-that-can-be-uploaded-to-a-texture, so add that abstraction.

### CODE REVIEW

**Comment by RAOF** [comment]
I am immensely skeptical that this breaks `mir_unit_tests.ProtobufSurfaceCounter` *only* on clang-16.04. Let's give it a retrigger…

**Comment by gerboland** [comment]
QtMir currently uses [mir:graphics::Buffer](https://github.com/ubports/qtmir/blob/master/src/modules/Unity/Application/mirbuffersgtexture.cpp) to fetch the texture, and relies on Qt's own shaders.

All it really wants is texture dimensions, and a texture id. Could Texture expose this?

Mind, Qt's own shaders won't deal with YUV textures, but that'll be QtMir's problem...

File: include/platform/mir/graphics/program.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+namespace mir
+{
+namespace graphics
+{
+namespace gl
+{
+class Program
+{
+public:
+    virtual ~Program();
```
Comment (AlanGriffiths) [inline include/platform/mir/graphics/program.h]: I'm curious: a CopyAssignable class with a virtual destructor? Can that ever be useful?

File: src/platform/symbols.map [hunk 2]
```diff
@@ -199,5 +199,18 @@ MIR_PLATFORM_0.32.3 {
...
+    vtable?for?mir::graphics::gl::Texture;
+
+    mir::graphics::gl::Program::?Program*;
+    typeinfo?for?mir::graphics::gl::Program;
+    vtable?for?mir::graphics::gl::Program;
+
+    mir::graphics::gl::ProgramFactory::?ProgramFactory*;
+    mir::graphics::gl::ProgramFactory::compile_fragment_shader*;
+    typeinfo?for?mir::graphics::gl::ProgramFactory;
+    vtable?for?mir::graphics::gl::ProgramFactory;
```
Comment (AlanGriffiths) [inline src/platform/symbols.map]: These don't belong in the MIR_PLATFORM_0.32.3 stanza. MIR_PLATFORM_1.1.0?

File: include/platform/mir/graphics/program.h (hunk 1 above)
Comment (AlanGriffiths) [reply include/platform/mir/graphics/program.h]: Reading further it looks as though CopyAssign shouldn't be enabled.

**Comment by RAOF** [comment]
Texture *could* expose this, however…
 
> Mind, Qt's own shaders won't deal with YUV textures, but that'll be QtMir's problem...

Qt's shaders *also* won't deal with textures from the `eglstream-kms` platform, which require an GLSL extension to be enabled and use a `samplerExternalOES` rather than `sampler2D`. I had *hoped* that the build-your-own-shader-with-string-interpolation thing would be flexible enough, but if Qt doesn't let you access the shaders its going to use *at all* that'll be a problem.

**Comment by RAOF** [comment]
Hm, and now a failure on `mir_unit_tests.ProtobufSocketMultiClientCommunicator`, 18.04-clang?

**Comment by gerboland** [comment]
> Qt's shaders _also_ won't deal with textures from the `eglstream-kms` platform, which require an GLSL extension to be enabled and use a `samplerExternalOES` rather than `sampler2D`. I had _hoped_ that the build-your-own-shader-with-string-interpolation thing would be flexible enough, but if Qt doesn't let you access the shaders its going to use _at all_ that'll be a problem.

Ok, then as we discussed on IRC, I expect QtMir can support this by implementing a custom QSGGraphicsNode, using the Mir-provided shader as the QSGMaterial. Would be good to check this works before committing

**Comment by AlanGriffiths** [comment]
Still fine with me.

bors r+

**Pull request merged**